In [4]:
a=int(input("Enter 1st number:"))
b=int(input("Enter 2nd number:"))

if (a>b):
    print(f"1st number ",a,"is greater than 2nd number ",b)
elif (b>a):
    print(f"2nd number ",b,"is greater than 1st number ",a)
else:
    print(f"1st number  ",a," and 2nd number ", b," are equal")


Enter 1st number: 99
Enter 2nd number: 99


1st number   99  and 2nd number  99  are equal


#List declaration
day_list=["Sunday","Monday"]
print(f"day_ist[]:",day_list)

#extend will add the element to end of list
day_list.extend(["Wednesday","Thursday"])

print(f"day_ist[]:",day_list)

print(f"length of day_ist[]:", len(day_list))
print(f"day_list[0]", day_list[0]) #first element in list will be displayed
#another way to print 1st element, find the length and multiply with -1
length=len(day_list)
length=length * -1
print(f"day_list[0]", day_list[length])


#append will take argument and it will get added the end
day_list.append("Friday")
print(f"day_ist[]:",day_list)

day_list.append("Saturady")
print(f"day_ist[]:",day_list)

#insert() method takes position and value to the list
day_list.insert(2,"Tuesday")
print(f"day_ist[]:",day_list)

#remove method can remove any data element from the list
day_list.remove(day_list[0])
print(f"day_list[]:",day_list)

#pop() - will remove last element in the list
day_list.pop()
print(f"day_list[]:",day_list)

#index() - will provide 1st occurrence of the element,
# if element does not exist it will through exception
try:
    print(f"Position of Friday is:",day_list.index("Friday"))
except ValueError as e:
    print("Friday is not in the list",e)

try:
    print(f"Position of Bhaskar is:",day_list.index("Bhaskar"))
except ValueError as e:
    print("Bhaskar is not in the list",e)

#count() method returns number of occurrence of an element in the list
day_list.append("Friday")
day_list.append("Friday")
print(f"day_list[]:",day_list)
print(f"Number of occurrence of Friday ",day_list.count("Friday"))

#if element does not exists then count() method will return value zero
print(f"Number of occurrence of Bhaskar ",day_list.count("Bhaskar"))


#reverse () method will reverse the content of the list
print(f"day_list[]:",day_list)

#traversing each element
for i in range(0,len(day_list)):
    print(f"day_ist[{i}]",day_list[i])

In [1]:
print ("Hello world")

Hello world


#Syntax

In [6]:
import sys
sys.version

'3.13.9 | packaged by Anaconda, Inc. | (main, Oct 21 2025, 19:09:58) [MSC v.1929 64 bit (AMD64)]'

In [7]:
import math
math.factorial(5)

120

In [16]:
print ('BSE NSE')
#print ("National Stock exchange")
print("BSE # NSE")
print ("BSE Mumbai")

usd="Dollar $"; inr="Rupee \u20b9"
print ("USD:", usd) ; print ("INR:", inr)

paragraph=""" This to demonstarte the multiple lines
display like paragrpah
"""

print (paragraph)

#multiline comments is achiveibale using three # or '

""" This program 
meant for adding two numbers
"""


BSE NSE
BSE # NSE
BSE Mumbai
USD: Dollar $
INR: Rupee ₹
 This to demonstarte the multiple lines
display like paragrpah



' This program \nmeant for adding two number\n'

In [19]:
import keyword
print (keyword.kwlist)

['False', 'None', 'True', 'and', 'as', 'assert', 'async', 'await', 'break', 'class', 'continue', 'def', 'del', 'elif', 'else', 'except', 'finally', 'for', 'from', 'global', 'if', 'import', 'in', 'is', 'lambda', 'nonlocal', 'not', 'or', 'pass', 'raise', 'return', 'try', 'while', 'with', 'yield']


In [21]:
print (dir(__builtins__))

['ArithmeticError', 'AssertionError', 'AttributeError', 'BaseException', 'BaseExceptionGroup', 'BlockingIOError', 'BrokenPipeError', 'BufferError', 'BytesWarning', 'ChildProcessError', 'ConnectionAbortedError', 'ConnectionError', 'ConnectionRefusedError', 'ConnectionResetError', 'DeprecationWarning', 'EOFError', 'Ellipsis', 'EncodingWarning', 'EnvironmentError', 'Exception', 'ExceptionGroup', 'False', 'FileExistsError', 'FileNotFoundError', 'FloatingPointError', 'FutureWarning', 'GeneratorExit', 'IOError', 'ImportError', 'ImportWarning', 'IndentationError', 'IndexError', 'InterruptedError', 'IsADirectoryError', 'KeyError', 'KeyboardInterrupt', 'LookupError', 'MemoryError', 'ModuleNotFoundError', 'NameError', 'None', 'NotADirectoryError', 'NotImplemented', 'NotImplementedError', 'OSError', 'OverflowError', 'PendingDeprecationWarning', 'PermissionError', 'ProcessLookupError', 'PythonFinalizationError', 'RecursionError', 'ReferenceError', 'ResourceWarning', 'RuntimeError', 'RuntimeWarning

In [22]:
print (dir(keyword))


['__all__', '__builtins__', '__cached__', '__doc__', '__file__', '__loader__', '__name__', '__package__', '__spec__', 'iskeyword', 'issoftkeyword', 'kwlist', 'softkwlist']


In [23]:
import keyword
print (keyword.softkwlist)

['_', 'case', 'match', 'type']


In [33]:
sensor=1
if sensor==0:
    print("123")
else:
    print(456)

""" this progream for demostarte indentation
    """

456


' this progream for demostarte indentation\n    '

Assignment

In [34]:
a=0
print(type(a))
a="String"
print(type(a))

<class 'int'>
<class 'str'>


In [35]:
print=20
print(print)

TypeError: 'int' object is not callable

In [1]:
print=20
print("Hello")

TypeError: 'int' object is not callable

In [1]:
print("Hello")

Hello


In [3]:
type(20)


int

In [4]:
type("hello")


str

In [5]:
type(print)

builtin_function_or_method

In [6]:
print (type(print))

<class 'builtin_function_or_method'>


In [7]:
print (type(type))

<class 'type'>


In [8]:
#id fucntion'

a="hello"

print(id(a))

1426451250512


In [9]:
a=20
b=20
print(a,b)
print (id(a),id(b))

20 20
140719711102472 140719711102472


In [10]:
#int values -5 to 256 are prefixed in memory

a=-5
b=-5
print(a,b)
print (id(a),id(b))



-5 -5
140719711101672 140719711101672


In [15]:
#positionla mapping

v1,v2,v3=10,20.4,"Hello"

print ("v1:",v1);print ("v2:",v2);print ("v3:",v3)

v1: 10
v2: 20.4
v3: Hello


In [16]:
#Keyboard Input

sub=input("Enter subject")
print(type(sub))
print (sub)


Enter subject pds


<class 'str'>
pds


In [18]:
#input fucntion always returns str

marks=input("Enter marks:")
print(marks+20)


Enter marks: 12


32


In [19]:
#input fucntion always returns str

marks=input("Enter marks:")
print(int(marks)+20)

Enter marks: 45


65


In [20]:
not(True)


False

In [21]:
not(False)

True

In [22]:
not(bool(5))

False

In [23]:
not(bool(-1))

False

In [24]:
not(bool(0))

True

In [30]:
temp=input("Enter temprature:")
temp=int(temp)

if temp>99:
    print("it is hot")
elif temp<60:
     print("it is cold")
elif temp>=60 and temp<=99:
     print("it is just fine")
    



Enter temprature: 59


it is cold


In [34]:
def addfun(num1,num2):
    output=num1+num2
    print(id(output),output)
    return output
    
x=addfun(5667,2000)

print(id(x),x)

1426507265456 7667
1426507265456 7667


###########

String

In [36]:
import os
os.getcwd()

'C:\\Users\\dhruv'

In [37]:
a,b,c = 20, 23.45, 23.4j

print(type(a))
print(type(b))
print(type(c))

<class 'int'>
<class 'float'>
<class 'complex'>


In [38]:
#Membership operator in

'good' in 'this is a good book'


True

In [39]:
'good' not in 'this is a good book'

False

In [40]:
"p".upper() in "pds".upper()

True

In [41]:
print(dir(str))


['__add__', '__class__', '__contains__', '__delattr__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getitem__', '__getnewargs__', '__getstate__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__iter__', '__le__', '__len__', '__lt__', '__mod__', '__mul__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__rmod__', '__rmul__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', 'capitalize', 'casefold', 'center', 'count', 'encode', 'endswith', 'expandtabs', 'find', 'format', 'format_map', 'index', 'isalnum', 'isalpha', 'isascii', 'isdecimal', 'isdigit', 'isidentifier', 'islower', 'isnumeric', 'isprintable', 'isspace', 'istitle', 'isupper', 'join', 'ljust', 'lower', 'lstrip', 'maketrans', 'partition', 'removeprefix', 'removesuffix', 'replace', 'rfind', 'rindex', 'rjust', 'rpartition', 'rsplit', 'rstrip', 'split', 'splitlines', 'startswith', 'strip', 'swapcase', 'title', 'translate', 'upper', 'zfill']


In [42]:
help(str.replace)


Help on method_descriptor:

replace(self, old, new, /, count=-1) unbound builtins.str method
    Return a copy with all occurrences of substring old replaced by new.

      count
        Maximum number of occurrences to replace.
        -1 (the default value) means replace all occurrences.

    If the optional argument count is given, only the first count occurrences are
    replaced.



In [1]:
help(os)


NameError: name 'os' is not defined

In [44]:
print('hh')

hh


In [2]:
ms1="king ring, SING"
ns1=ms1.replace("ng","hh")
print("ms1",ms1); print ("ns1",ns1)

ms1 king ring, SING
ns1 kihh rihh, SING


In [47]:
ms1="king ring, SING"
ns1=ms1.replace("ng","hh",-1)
print("ms1",ms1); print ("ns1",ns1)

ms1 king ring, SING
ns1 kihh rihh, SING


In [48]:
ms1="king ring, SING"
ns1=ms1.replace("ng","hh",-2)
print("ms1",ms1); print ("ns1",ns1)

ms1 king ring, SING
ns1 kihh rihh, SING


In [3]:
help(str.find)


Help on method_descriptor:

find(self, sub[, start[, end]], /) unbound builtins.str method
    Return the lowest index in S where substring sub is found, such that sub is contained within S[start:end].

    Optional arguments start and end are interpreted as in slice notation.
    Return -1 on failure.



In [51]:
help(str.rfind)

Help on method_descriptor:

rfind(self, sub[, start[, end]], /) unbound builtins.str method
    Return the highest index in S where substring sub is found, such that sub is contained within S[start:end].

    Optional arguments start and end are interpreted as in slice notation.
    Return -1 on failure.



In [59]:
## enumerate fucntion provide index and value
ms1="Nationla Stock Exchane"
print(list(enumerate(ms1)))
print(enumerate(ms1))
      

[(0, 'N'), (1, 'a'), (2, 't'), (3, 'i'), (4, 'o'), (5, 'n'), (6, 'l'), (7, 'a'), (8, ' '), (9, 'S'), (10, 't'), (11, 'o'), (12, 'c'), (13, 'k'), (14, ' '), (15, 'E'), (16, 'x'), (17, 'c'), (18, 'h'), (19, 'a'), (20, 'n'), (21, 'e')]


In [66]:
age=input("Enter your age:")

if age.isdigit():
    age1=int (age)
    if age1>0:
        print("Valid age:",age1)
else:
    print("Invalid age",age)

Enter your age: 23h


Invalid age 23h


In [69]:
mylist = ["Bhaskar","Kuber","Suresh"]
sep=" "
sep.join(mylist)
         

'Bhaskar Kuber Suresh'

In [70]:
help(str.join)


Help on method_descriptor:

join(self, iterable, /) unbound builtins.str method
    Concatenate any number of strings.

    The string whose method is called is inserted in between each given string.
    The result is returned as a new string.

    Example: '.'.join(['ab', 'pq', 'rs']) -> 'ab.pq.rs'



In [11]:
#Avrerga e=

num=50
sub="PDS"
avg=89.95
#unconstraint
print ("Average marks of {} Students in {} is {}".format(num,sub,avg))
#data type constraint placehodler
print ("Average marks of {:d} Students in {:s} is {:f}".format(num,sub,avg))
print ("Average marks of {:d} Students in {:s} is {:.2f}".format(num,sub,avg))
#print ("Average marks of {:d} Students in {:s} is {:.2f}".format(sub,num,avg))

#replacement index contraint
print ("Average marks of {0} Students in {1} is {2}".format(num,sub,avg))
print ("Average marks of {0} Students in {0} is {0}".format(num,sub,avg))

#print ("Average marks of {0} Students in {1} is {2}".format(num,sub))
print ("Average marks of {0:d} Students in {1:s} is {2:.2f}".format(num,sub,avg))

Average marks of 50 Students in PDS is 89.95
Average marks of 50 Students in PDS is 89.950000
Average marks of 50 Students in PDS is 89.95
Average marks of 50 Students in PDS is 89.95
Average marks of 50 Students in 50 is 50
Average marks of 50 Students in PDS is 89.95


In [17]:
#DIY: Create a formatted strng to demo, left align, right align and center align
## <, >, and ^

s1="jdghjhdgfjdjfgjdgfj"
#print("Left align:",left_align(s1))


In [18]:
m1=[11,22,33]
print(type(m1))
print(len(m1))
print(m1)

<class 'list'>
3
[11, 22, 33]


In [19]:
m1[0]=232323
print(type(m1))
print(len(m1))
print(m1)

<class 'list'>
3
[232323, 22, 33]


In [25]:
m2=[11,22,"pds",[1212,"fdfd",23343,"dfdg"]]
print(m2[-1][-1][-2])    

d


In [30]:
matrix43=[ [12,13,14,15],[1,2,3,4],[10,20,30,40]]
print(matrix43[1][1])

print(matrix43[-1][-1])
print(20 in matrix43[-1])
print([1,2,3,4] in matrix43)

2
40
True
True


In [32]:
m1=[1,2,3]
m2=m1.append("gfhjsdgfh") # inplace update .. self update similar to pointer in C
print(m2)
print(m1)
m2=m1
print(m2)

None
[1, 2, 3, 'gfhjsdgfh']
[1, 2, 3, 'gfhjsdgfh']


In [35]:
mylist1=list("hdgfdfdfdsfh")
print(mylist1)


['h', 'd', 'g', 'f', 'd', 'f', 'd', 'f', 'd', 's', 'f', 'h']


In [36]:
m1=list("nsebsemumbai")
print(m1)
while 'e' in m1:
    m1.remove("e")
    print(m1)


['n', 's', 'e', 'b', 's', 'e', 'm', 'u', 'm', 'b', 'a', 'i']
['n', 's', 'b', 's', 'e', 'm', 'u', 'm', 'b', 'a', 'i']
['n', 's', 'b', 's', 'm', 'u', 'm', 'b', 'a', 'i']


In [45]:
ord("i") #ordinal value retunr ASCII

105

In [49]:
i=100
chr(i) 
#return charcter value of ASCII value

'd'

In [54]:
i=1
while i in range(1,256):
    print(chr(i), "\t",ord(chr(i)))
    i+=1

 	 1
 	 2
 	 3
 	 4
 	 5
 	 6
 	 7
 	 8
	 	 9

 	 10
 	 11
 	 12
 	 13
 	 14
 	 15
 	 16
 	 17
 	 18
 	 19
 	 20
 	 21
 	 22
 	 23
 	 24
 	 25
 	 26
 	 27
 	 28
 	 29
 	 30
 	 31
  	 32
! 	 33
" 	 34
# 	 35
$ 	 36
% 	 37
& 	 38
' 	 39
( 	 40
) 	 41
* 	 42
+ 	 43
, 	 44
- 	 45
. 	 46
/ 	 47
0 	 48
1 	 49
2 	 50
3 	 51
4 	 52
5 	 53
6 	 54
7 	 55
8 	 56
9 	 57
: 	 58
; 	 59
< 	 60
= 	 61
> 	 62
? 	 63
@ 	 64
A 	 65
B 	 66
C 	 67
D 	 68
E 	 69
F 	 70
G 	 71
H 	 72
I 	 73
J 	 74
K 	 75
L 	 76
M 	 77
N 	 78
O 	 79
P 	 80
Q 	 81
R 	 82
S 	 83
T 	 84
U 	 85
V 	 86
W 	 87
X 	 88
Y 	 89
Z 	 90
[ 	 91
\ 	 92
] 	 93
^ 	 94
_ 	 95
` 	 96
a 	 97
b 	 98
c 	 99
d 	 100
e 	 101
f 	 102
g 	 103
h 	 104
i 	 105
j 	 106
k 	 107
l 	 108
m 	 109
n 	 110
o 	 111
p 	 112
q 	 113
r 	 114
s 	 115
t 	 116
u 	 117
v 	 118
w 	 119
x 	 120
y 	 121
z 	 122
{ 	 123
| 	 124
} 	 125
~ 	 126
 	 127
€ 	 128
 	 129
‚ 	 130
ƒ 	 131
„ 	 132
… 	 133
† 	 134
‡ 	 135
ˆ 	 136
‰ 	 137
Š 	 138
‹ 	 13

In [ ]:
#DIY what is diff bm shallo copy and deep copy
#work with copy module, .cop and .deepcopy



In [55]:
#touple

days=(0,1,2,3,4,5,6)

daysname=("sunday","monday","tuesday","wednesday","thursday","friday","saturday")

print(days)
print(daysname)


(0, 1, 2, 3, 4, 5, 6)
('sunday', 'monday', 'tuesday', 'wednesday', 'thursday', 'friday', 'saturday')


In [59]:
print(type(days))
dummy=("sdfdf",)
print(type(dummy))

<class 'tuple'>
<class 'tuple'>


In [69]:
mt1=(11,22,"Hello",[66,77,88])
print(mt1)
print(mt1[2])
print(mt1[2][0])
#mt1[2][0]='j'
print(mt1[2][0])

(11, 22, 'Hello', [66, 77, 88])
Hello
H
H


In [71]:
mt1=(11,22,"Hello",[66,77,88])
print(mt1)
print(mt1[2])
print(mt1[2][0])
#mt1[3][0]='j'
#print(mt1)

#adding an element to touple
list1 = list(mt1)
list1.append("KSB")
mt1=tuple(list1)
print(mt1)

(11, 22, 'Hello', [66, 77, 88])
Hello
H
(11, 22, 'Hello', [66, 77, 88], 'KSB')


In [72]:
mt1=(11,22,"Hello",[66,77,88])
print(mt1)
print(mt1[2])
print(mt1[2][0])

#adding an element to touple
list1 = list(mt1)
list1.append("KSB")
mt1=tuple(list1)
print(mt1)

(11, 22, 'Hello', [66, 77, 88])
Hello
H
(11, 22, 'Hello', [66, 77, 88], 'KSB')


In [2]:
## Dictionary

dict1={"Bank":"SBI","Branch":"M G Road","City":"Mumbai"}

print(type(dict1))
print(len(dict1))
l1=len(dict1)

i=0
while i in (0,l1):
    print(f"dict1[0]",dict1["Bank"])
    i+=1

print(dict1.pop("City"))
print(dict1)

<class 'dict'>
3
dict1[0] SBI
Mumbai
{'Bank': 'SBI', 'Branch': 'M G Road'}


In [5]:
dict1.setdefault("State","MH")
print(dict1)

{'Bank': 'SBI', 'Branch': 'M G Road', 'State': 'MH'}


In [6]:
dict1={"Bank":"SBI","Branch":"M G Road","City":"Mumbai"}
dict2={"State":"KA","Bank":"Axis"}
print(dict1)
print(dict2)

dict1.update(dict2)
print(dict1)
print(dict2)


{'Bank': 'SBI', 'Branch': 'M G Road', 'City': 'Mumbai'}
{'State': 'KA', 'Bank': 'Axis'}
{'Bank': 'Axis', 'Branch': 'M G Road', 'City': 'Mumbai', 'State': 'KA'}
{'State': 'KA', 'Bank': 'Axis'}


In [8]:
## Iterator
klist=["Bank","Branch","City"]
vlist=["SBI","M G Road","Mumbai"]

zobj=zip(klist,vlist)
print(zobj)
print(list(zobj))
print(list(zobj))

[('Bank', 'SBI'), ('Branch', 'M G Road'), ('City', 'Mumbai')]
[]


In [9]:
klist=["Bank","Branch","City"]
vlist=["SBI","M G Road","Mumbai"]

zobj=zip(klist,vlist)
md1=dict(zobj)
print(md1)


{'Bank': 'SBI', 'Branch': 'M G Road', 'City': 'Mumbai'}


#########SET


In [16]:
s1={1,2,3,4,5,3}
print(type(s1))
print(f"s1",s1)
print(len(s1))
s2={5,6,7,8,9,10,11}
print(f"s2",s2)
s3=s1.union(s2)
print("Union",s3)

s3=s1.intersection(s2)
print("Intersection",s3)


<class 'set'>
s1 {1, 2, 3, 4, 5}
5
s2 {5, 6, 7, 8, 9, 10, 11}
Union {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11}
Intersection {5}


In [17]:
print(dir(set))

['__and__', '__class__', '__class_getitem__', '__contains__', '__delattr__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getstate__', '__gt__', '__hash__', '__iand__', '__init__', '__init_subclass__', '__ior__', '__isub__', '__iter__', '__ixor__', '__le__', '__len__', '__lt__', '__ne__', '__new__', '__or__', '__rand__', '__reduce__', '__reduce_ex__', '__repr__', '__ror__', '__rsub__', '__rxor__', '__setattr__', '__sizeof__', '__str__', '__sub__', '__subclasshook__', '__xor__', 'add', 'clear', 'copy', 'difference', 'difference_update', 'discard', 'intersection', 'intersection_update', 'isdisjoint', 'issubset', 'issuperset', 'pop', 'remove', 'symmetric_difference', 'symmetric_difference_update', 'union', 'update']


In [1]:
import os
os.getcwd()


'C:\\Users\\dhruv'

In [4]:
import sys
sys.version

'3.13.9 | packaged by Anaconda, Inc. | (main, Oct 21 2025, 19:09:58) [MSC v.1929 64 bit (AMD64)]'

In [9]:
str="bhaskar bhaskar OO hhsh, sh"
str.format(str,"bhas","sh")
print(str)

bhaskar bhaskar OO hhsh, sh


//08-Aug-2026
decision making

In [12]:
t1=999
t2=98
climate1="hot" if t1>99 else "not hot"
print(t1,climate1)

climate1="hot" if t2>99 else "not hot"
print(t2,climate1)

999 hot
999 not hot


In [14]:
t1=23

result="even" if (t1%2==0) else "odd"

print (t1,result)

23 odd


In [17]:
age=35

if age>18:
    if age>=80:
        tax="Senior Citizen"
    elif age>=60:
        tax="Young Senior Citizen"
    else:
        tax="Regular"
else:
    tax="Minor"

print(age,tax)
    

35 Regular


In [20]:
age=60
tax="Minor" if age<18 else "SSC" if age>80 else "YSC" if age > 60 else "Regular"
print(age,tax)

79 YSC


#Range

In [22]:
print(list(range(6)))

print(list(range(10,20)))

list(range(10,20,2))

[0, 1, 2, 3, 4, 5]
[10, 11, 12, 13, 14, 15, 16, 17, 18, 19]


[10, 12, 14, 16, 18]

In [23]:
import sys
m1=list(range(100))
sys.getsizeof(m1)

856

In [27]:
r1=range(1000) #start, end and incremental value, hence moemiry of range object will always constant 
print(sys.getsizeof(r1))
m1=list(r1)
print(sys.getsizeof(m1))

print(r1[0],r1[-1])

print(m1[0],m1[-1])

r1=range(10) #start, end and incremental value
print(sys.getsizeof(r1))
m1=list(r1)
print(sys.getsizeof(m1))

print(r1[0],r1[-1])

print(m1[0],m1[-1])

48
8056
0 999
0 999
48
136
0 9
0 9


In [29]:
#list(range(10.1,10.5,0.1))

In [31]:
#list(range('a','b','c'))

In [3]:
lcount=10
while lcount>0:
    print(lcount)
    lcount=lcount-1
    

    


10
9
8
7
6
5
4
3
2
1


In [5]:
m=10
while m>=30:
    print(m)
    m=m+3

In [3]:
print(dir(__builtin__))

['ArithmeticError', 'AssertionError', 'AttributeError', 'BaseException', 'BaseExceptionGroup', 'BlockingIOError', 'BrokenPipeError', 'BufferError', 'BytesWarning', 'ChildProcessError', 'ConnectionAbortedError', 'ConnectionError', 'ConnectionRefusedError', 'ConnectionResetError', 'DeprecationWarning', 'EOFError', 'Ellipsis', 'EncodingWarning', 'EnvironmentError', 'Exception', 'ExceptionGroup', 'False', 'FileExistsError', 'FileNotFoundError', 'FloatingPointError', 'FutureWarning', 'GeneratorExit', 'IOError', 'ImportError', 'ImportWarning', 'IndentationError', 'IndexError', 'InterruptedError', 'IsADirectoryError', 'KeyError', 'KeyboardInterrupt', 'LookupError', 'MemoryError', 'ModuleNotFoundError', 'NameError', 'None', 'NotADirectoryError', 'NotImplemented', 'NotImplementedError', 'OSError', 'OverflowError', 'PendingDeprecationWarning', 'PermissionError', 'ProcessLookupError', 'PythonFinalizationError', 'RecursionError', 'ReferenceError', 'ResourceWarning', 'RuntimeError', 'RuntimeWarning

In [7]:
for char in "pythonpds":
    print(char)

p
y
t
h
o
n
p
d
s


In [9]:
for var in "dataScience":
    if var in "aeiouAEIOU":
        print(var)

a
a
i
e
e


In [10]:
for item in [10,20,"hello"]:
    print(item)


10
20
hello


In [11]:
for i in range(20):
    print(i)

0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19


In [6]:
import time
print(time.time())
for i in range(100000000):
    j=i

print(time.time())


1786181670.342738
1786181677.1296885


In [7]:
import time
m1=list(range(100000000))
print(time.time())
for i in m1:
    j=i

print(time.time())

1786181691.4578114
1786181709.9079137


In [8]:
a=10,20,30
print(a)
print(type(a))

(10, 20, 30)
<class 'tuple'>


In [9]:
a,b=10,20
print(a,b)
print(type(a),type(b))

10 20
<class 'int'> <class 'int'>


In [12]:
a,b,c="pdd"
print(a)

p


In [3]:
str1="My bad bad , good  , dog mad mad bed bed good good god"

strList = list(str1.split())
md1={}
print(strList)
for str2 in str1.split():
    print(str2)
    md1[str2]=md1.setdefault(str2,0)+1

print(md1)


['My', 'bad', 'bad', ',', 'good', ',', 'dog', 'mad', 'mad', 'bed', 'bed', 'good', 'good', 'god']
My
bad
bad
,
good
,
dog
mad
mad
bed
bed
good
good
god
{'My': 1, 'bad': 2, ',': 2, 'good': 3, 'dog': 1, 'mad': 2, 'bed': 2, 'god': 1}


In [11]:
lcount=9
while(lcount>0):
    lcount-=1
    if lcount==6:
        continue

    print(lcount)
    

8
7
5
4
3
2
1
0


In [4]:
slist=[]
while True:
    str1=input("enter shop item:")
    if str1=="endshopping":
        break;
    slist.append(str1)

print(slist)

enter shop item: nbvsdnf
enter shop item: nbvsdnf
enter shop item: nbvsdf
enter shop item: endshopping


['nbvsdnf', 'nbvsdnf', 'nbvsdf']


In [7]:
def add(i,j):
    return (i+j)
    
m1=add(10,20)
print(m1)




30


In [8]:
print(add(100,200))

300


In [2]:
#Rule 1: While defining a function, all required params must be specified before the default_params
#positional order passing
#parameter-keyword based passing

def mf1(a):
    print(a**2)

def mf1(a,b=1):
    print(a**b)

mf1(5)
mf1(5,3)
mf1(3,5)

5
125
243


In [ ]:
#rule2 : While calling a function positional mapped values must be specified befoe parameter keyword mapped values

#if we put * value infront of parameter
#Rule 3: A vairable tuple can be pouplated only in positional mapping style and fucntion can have only one variable tuple

#if we put two stars ** infront of parameter
#rule 4: A variable dictionary can be populated in positional mapping style, provided key value pair and it should be defines at end of the
#fucntion signature

def mf12(r1,r2,d1=111,d2=222,*args, **kwargs):
    print(r1,r2)
    print(d1,d2)
    print(args)
    print(kwargs)

#mf12(10,20,0,0,50,"Hello","world",7,7,k1=1111,k2=3333)


mf12(10,20,d1=7,d2=7,k1=1111,k2=3333)

In [ ]:
#LAMDA

In [37]:
def addsum1 (p1,p2):
    return p1+p2

x1=addsum1(10,20)
print(x1)
print(type(addsum1))


30
<class 'function'>


In [38]:
evenf1= lambda i : (i%2)==0

print(evenf1(20))
print(evenf1(21))

True
False


In [41]:
evenf2= lambda i :  "even" if (i%2==0) else "odd"

print(evenf1(20))
print(evenf1(21))

True
False


In [42]:
kmlist=[1,2,3]
mlist=[]

for km in kmlist:
    mlist.append(km*1000)

print(mlist)
    

[1000, 2000, 3000]


In [5]:
#map method - transformation fucntion
kmf=lambda km:km*1000
kmlist=[1,2,3]

mobj=map(kmf,kmlist)
print(type(mobj))
print(list(mobj))

kmlist=[4,5,6]
mobj1=map(lambda km:km*1000,kmlist)
print(type(mobj1))
print(list(mobj1))

<class 'map'>
[1000, 2000, 3000]
<class 'map'>
[4000, 5000, 6000]


###09-Aug-2026

In [9]:
#filter

m1=list('data science')
print(m1)
fobj=filter(lambda char: char in "aeiouAEIOU",m1)
print(type(fobj))
print(list(fobj))

['d', 'a', 't', 'a', ' ', 's', 'c', 'i', 'e', 'n', 'c', 'e']
<class 'filter'>
['a', 'a', 'i', 'e', 'e']


In [11]:
import functools as ft
help(ft.reduce)

Help on built-in function reduce in module _functools:

reduce(...)
    reduce(function, iterable[, initial], /) -> value

    Apply a function of two arguments cumulatively to the items of an iterable, from left to right.

    This effectively reduces the iterable to a single value.  If initial is present,
    it is placed before the items of the iterable in the calculation, and serves as
    a default when the iterable is empty.

    For example, reduce(lambda x, y: x+y, [1, 2, 3, 4, 5])
    calculates ((((1 + 2) + 3) + 4) + 5).



In [13]:
ft.reduce(lambda x,y:x+y,[5,10,15,20,25])

75

In [14]:
m1=[33,44,11,77,22,55]
ft.reduce(lambda x,y:x if x>y else y,m1)

77

In [1]:
#DIY In a fucntion signature, you pass "*" or "/"?

def doit(m1,m2,/):
    return m1+m2

print(doit(12,13))

25


In [26]:
m1=[2,3,4,5,6]
m2=[ i**2 for i in m1]
print(m2)

[4, 9, 16, 25, 36]


In [29]:
m1=[102,103,100,99,98.4]
m2=[ round((i-32)*5/9,3) for i in m1]
print(m2)

[38.889, 39.444, 37.778, 37.222, 36.889]


In [33]:
m1=list('data science')
m2=[ char for char in m1 if char in "aeiouAEIOU" ]
print(m2)

['a', 'a', 'i', 'e', 'e']


In [37]:
m1=[10,11,12,13,14]
m2=[ num**2 for num in m1 if num%2==0]
print(m2)

[100, 144, 196]


In [51]:
m2=[]
for i in range(1,31):
    if(i%3==0):
        if(i%2==0):
            t1=(i,i**2)
        else:
            t1=(i,i**3)
        m2.append(t1)
print(m2)

[(3, 27), (6, 36), (9, 729), (12, 144), (15, 3375), (18, 324), (21, 9261), (24, 576), (27, 19683), (30, 900)]


In [55]:
ulist=['a1','b1','c1','d1']
plist=['i1','j1','k1','m1']

m2=[ (u,p) for u in ulist if u != "b1" for p in plist if p != "j1"]
print(m2)

[('a1', 'i1'), ('a1', 'k1'), ('a1', 'm1'), ('c1', 'i1'), ('c1', 'k1'), ('c1', 'm1'), ('d1', 'i1'), ('d1', 'k1'), ('d1', 'm1')]


In [64]:
#fibannaci series using recursive fucntionabs
fseries=[0,1,1]
def fibo(a,b,count):
    if(count>35):
        return fseries
    c=a+b
    fseries.append(c)
    count=count+1
    return(fibo(b,c,count))

print(fibo(1,1,3))
    

[0, 1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 144, 233, 377, 610, 987, 1597, 2584, 4181, 6765, 10946, 17711, 28657, 46368, 75025, 121393, 196418, 317811, 514229, 832040, 1346269, 2178309, 3524578, 5702887, 9227465]


In [ ]:
globals()

Scope of Variables

In [67]:
'hh1' in globals()

True

In [66]:
hh1=12


In [70]:
g1 =11
g2=22

globals()
def mf99(v3):
    global g1
    g1=33
    g2=55
    print(locals())
    print("inside fucntion",g1,g2,v3)

print("outside function",g1,g2)
mf99(10)
print("outside function",g1,g2)
print(locals())

outside function 11 22
{'v3': 10, 'g2': 55}
inside fucntion 33 55 10
outside function 33 22


#LEGB 
#Local vaiablesabs
#Enclosing Scomplex
#Global Scomple
#Built-in


In [71]:
#Constrcuts

m1=[n%3 for n in range(20)]
print(m1)

[0, 1, 2, 0, 1, 2, 0, 1, 2, 0, 1, 2, 0, 1, 2, 0, 1, 2, 0, 1]


In [72]:
m1={n%3 for n in range(20)}
print(m1)

{0, 1, 2}


In [75]:
m1=list('datascience')
print (m1)

md1={ char:m1.count(char) for char in m1}
print(md1)

['d', 'a', 't', 'a', 's', 'c', 'i', 'e', 'n', 'c', 'e']
{'d': 1, 'a': 2, 't': 1, 's': 1, 'c': 2, 'i': 1, 'e': 2, 'n': 1}


In [76]:
def mf888():
    pass #dummy, filler statement

In [88]:
m1=[ i*j for i in range(2,5) for j in range(1,5)]
print(m1)

[2, 4, 6, 8, 3, 6, 9, 12, 4, 8, 12, 16]


In [89]:
import numpy as np
